rfc/mediachain-rfc-1.md

```diff
@@ -0,0 +1,432 @@
+ }
+
+EntityLinkCell = {
+ "type" : "entityLink"
+ "chain" : <Reference>
+ "entity" : <Reference>
+ "entityLink" : <Reference>
+ ["relationship" : <Reference>]
+ "signatures" : <Signatures>
+ <Key> : <Value> ... ; entity relationship metadata
+ }
+
+ArtefactChainCell =
+   <ArtefactUpdateCell>
+ | <ArtefactCreationCell>
+ | <ArtefactDerivationCell>
+ | <ArtefactOwnershipCell>
+ | <ArtefactReferenceCell>
+ | <Nil>
+ 
+ArtefactUpdateCell = {
+ "type" : "artefactUpdate"
+ "chain" : <Reference>
+ "artefact" : <Reference>
+ "signatures" : <Signatures>
+ <Key> : <Value> ... ; metadata updates
+ }
+
+ArtefactCreationCell = {
+ "type" : "artefactCreatedBy"
+ "chain" : <Reference>
+ "artefact" : <Reference>
+ "entity" : <Reference>
+ "signatures" : <Signatures>
+ <Key> : <Value> ... ; creation metadata
+ }
+
+ArtefactDerivationCell = {
+ "type" : "artefactDerivedBy"
+ "chain" : <Reference>
+ "artefact" : <Reference>
+ "artefactOrigin" : <Reference>
+ "signatures" : <Signatures>
+ <Key> : <Value> ... ; creation metadata
+ }
+
+ArtefactOwnershipCell = {
+ "type" : "artefactRightsOwnedBy"
+ "chain" : <Reference>
+ "artefact" : <Reference>
+ "entity" : <Reference>
+ "signatures" : <Signatures>
+ <Key> : <Value> ... ; IP ownership metadata
+ }
+
+ArtefactReferenceCell = {
+ "type" : "artefactReferencedBy"
+ "chain" : <Reference>
+ "artefact" : <Reference>
+ ["entity" : <Reference>]
+ ["url" : <URL>]
+ "signatures" : <Signatures>
+ <Key> : <Value> ... ; reference metadata
+  }
+```
+
+## Indexing and Querying the Mediachain
+
+### The Journal
+
+Mediachain data are persistent and available through the IPFS network.
+However they are not _discoverable_ without the help of an index which
+maintains references to entities, artefacts and chain heads.
+
+In order to make the mediachain indexable and discoverable we need to
+maintain a _Journal_ of updates to the data store. The journal is
+very similar to a blockchain in that it acts as a public ledger of all
+Mediachain data transactions in the system. By replaying the journal
+and fetching the data from IPFS, any node in the Internet can bootstrap
+an index that allows it to read and query the Mediachain.
+
+### Appending Data in the Mediachain
+
+The datastore initially is almost empty, containing only the `Nil` object
+as the bottom of all chains 
+For each entity and artefact added to the store, a corresponding entry
+is added to the journal, connecting a canonical reference with its chain
+pointing to the `Nil` object.
+Similarly, for every cell added to a chain, a corresponding entry is
+added to the journal, updating the head of the chain for a canonical
+reference.
+
+Thus the journal contents can be described as a sequence of entries
+with the following schema:
+```
+Journal = <JournalEntry> ...
+
+JournalEntry =
+ <CanonicalEntry>
+ <ChainEntry>
+
+CanonicalEntry = {
+ "type" = "insert"
+ "ref"  = <Reference> ; canonical reference for an entity or artefact
+ }
+ 
+ChainEntry = {
```

Comment:
Actually, on better thought I'll add timestamps -- I'll just wont specified the format for it.
When we have a better idea how the ledger protocol will work we can speicify it.